### PR TITLE
[ResourceList] Simplify / Improve interaction state logic

### PR DESCRIPTION
**Pull request by AndrewMusgrave**
### WHY are these changes introduced?

Iv'e seen this bug out a couple times in production. Simplifying the should also help make changes in the future

## 

### How to 🎩

Make sure to 🎩  with ⌨️ and 🖱 

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, Avatar, Card, ResourceList, FilterType, TextStyle} from '../src';

interface State {}

export default class ResourceListExample extends React.Component {
  state = {
    selectedItems: [],
    sortValue: 'DATE_MODIFIED_DESC',
    searchValue: '',
    appliedFilters: [
      {
        key: 'accountStatusFilter',
        value: 'Account enabled',
      },
    ],
  };

  handleSearchChange = (searchValue) => {
    this.setState({searchValue});
  };

  handleFiltersChange = (appliedFilters) => {
    this.setState({appliedFilters});
  };

  handleSortChange = (sortValue) => {
    this.setState({sortValue});
  };

  handleSelectionChange = (selectedItems) => {
    this.setState({selectedItems});
  };

  renderItem = (item) => {
    const {id, url, name, location, latestOrderUrl} = item;
    const media = <Avatar customer size="medium" name={name} />;
    const shortcutActions = latestOrderUrl
      ? [{content: 'View latest order', url: latestOrderUrl}]
      : null;
    return (
      <ResourceList.Item
        id={id}
        url={url}
        media={media}
        accessibilityLabel={`View details for ${name}`}
        shortcutActions={shortcutActions}
        persistActions
      >
        <h3>
          <TextStyle variation="strong">{name}</TextStyle>
        </h3>
        <div>{location}</div>
      </ResourceList.Item>
    );
  };

  render() {
    const resourceName = {
      singular: 'customer',
      plural: 'customers',
    };
    const items = [
      {
        id: 341,
        url: 'customers/341',
        name: 'Mae Jemison',
        location: 'Decatur, USA',
        latestOrderUrl: 'orders/1456',
      },
      {
        id: 256,
        name: 'Ellen Ochoa',
        location: 'Los Angeles, USA',
        latestOrderUrl: 'orders/1457',
      },
      {
        id: 3141,
        name: 'Mae Jemison',
        location: 'Decatur, USA',
        latestOrderUrl: 'orders/1456',
      },
      {
        id: 2576,
        url: 'customers/2576',
        name: 'Ellen Ochoa',
        location: 'Los Angeles, USA',
        latestOrderUrl: 'orders/1457',
      },
      {
        id: 323441,
        url: 'customers/323441',
        name: 'Mae Jemison',
        location: 'Decatur, USA',
        latestOrderUrl: 'orders/1456',
      },
      {
        id: 256786,
        url: 'customers/256786',
        name: 'Ellen Ochoa',
        location: 'Los Angeles, USA',
        latestOrderUrl: 'orders/1457',
      },
    ];
    const promotedBulkActions = [
      {
        content: 'Edit customers',
        onAction: () => console.log('Todo: implement bulk edit'),
      },
    ];
    const bulkActions = [
      {
        content: 'Add tags',
        onAction: () => console.log('Todo: implement bulk add tags'),
      },
      {
        content: 'Remove tags',
        onAction: () => console.log('Todo: implement bulk remove tags'),
      },
      {
        content: 'Delete customers',
        onAction: () => console.log('Todo: implement bulk delete'),
      },
    ];
    const filters = [
      {
        key: 'orderCountFilter',
        label: 'Number of orders',
        operatorText: 'is greater than',
        type: FilterType.TextField,
      },
      {
        key: 'accountStatusFilter',
        label: 'Account status',
        operatorText: 'is',
        type: FilterType.Select,
        options: ['Enabled', 'Invited', 'Not invited', 'Declined'],
      },
    ];
    const filterControl = (
      <ResourceList.FilterControl
        filters={filters}
        appliedFilters={this.state.appliedFilters}
        onFiltersChange={this.handleFiltersChange}
        searchValue={this.state.searchValue}
        onSearchChange={this.handleSearchChange}
        additionalAction={{
          content: 'Save',
          onAction: () => console.log('New filter saved'),
        }}
      />
    );
    return (
      <Card>
        <ResourceList
          resourceName={resourceName}
          items={items}
          renderItem={this.renderItem}
          selectedItems={this.state.selectedItems}
          onSelectionChange={this.handleSelectionChange}
          promotedBulkActions={promotedBulkActions}
          bulkActions={bulkActions}
          sortValue={this.state.sortValue}
          sortOptions={[
            {label: 'Newest update', value: 'DATE_MODIFIED_DESC'},
            {label: 'Oldest update', value: 'DATE_MODIFIED_ASC'},
          ]}
          onSortChange={(selected) => {
            this.setState({sortValue: selected});
            console.log(`Sort option changed to ${selected}.`);
          }}
          filterControl={filterControl}
        />
      </Card>
    );
  }
}
```

</details>

### CODE REVIEW

File: src/components/ResourceList/components/Item/Item.tsx [hunk 1]
```diff
@@ -68,6 +68,7 @@ export type CombinedProps =
       WithContextTypes<ResourceListContext>;
 
 const getUniqueCheckboxID = createUniqueIDFactory('ResourceListItemCheckbox');
+const ITEM_ID = 'ResourceList-Item-Overlay';
```
Comment (alex-page) [inline src/components/ResourceList/components/Item/Item.tsx]: Duplicate id's yikes

**Comment by AndrewMusgrave** [comment]
`id` should be unique now @alex-page . And there's a slight change in the focus state with mouse interactions since we never pass focus to the items, they shouldn't of had focus interaction states from the mouse. @sarahill would you mind giving this a 🎩? 😄

**Comment by sarahill** [comment]
The interaction states look good to me 👌 

I did notice some contrast issues with the link text on the interaction state backgrounds. Not something to fix in this issue but something I noticed.

[image](https://user-images.githubusercontent.com/8629173/56232726-726deb80-604f-11e9-8572-abf40b3d7f75.png)

**Comment by AndrewMusgrave** [comment]
@dleroux  These problems have to do with #1313 so I'll make changes there, rebase and let you know when 👌 

> Entering selectMode doesn't make the checkboxes appear.

I think I know what's causing this 🤞 

> Tabbing through, there's a few things that get focus that don't provide the visual feedback of what is focus.

This seems to be an outstanding issue occurring on master as well, I don't think we ever really designed small screen mode to work compatible with keyboard interactions (we probably should have). It should be (🤞) a small fix of not displaying certain elements while on screen screen. The issue is that even on small screen we still render the checkbox. So we tab into an item, it gets focus at the beginning then onto the checkbox, then tab into what visual on small screens.

> If the actions aren't persistent, then hovering makes the action disappear.

This is related to the issue above, actions shouldn't appear on small screens to begin with. Disappearing is the correct behaviour, we just shouldn't be displaying them on small screens.

> Also curious what was the original bug that brought on these changes?

I'm not sure 😅 We talked about it yesterday, you even saw the bug but we're not sure how to reproduce it 😭 I saw it a few times while working on multiselect and it's been bugging me since. Maybe events are't properly firing? 🤷‍♂️ Hoping with simplified logic and reduced events we won't see it anymore.

**Comment by AndrewMusgrave** [comment]
@dleroux Iv'e rebased with the base branch and pushed the actions fix to master so this should be 👌

**Pull request merged**